### PR TITLE
commits: use empty strings instead of placeholders for missing name or email

### PR DESCRIPTION
**Pull request by emilykfox**
https://github.com/martinvonz/jj/issues/2094. I went with `JJ_EMPTY_STRING` as the placeholder in `git`, because the word `PLACEHOLDER` seems redundant when you're already using SCREAMING_SNAKE_CASE.

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/template_builder.rs
Comment (martinvonz) [inline cli/src/template_builder.rs:1]: I wonder if it's better to make this change in `cli/src/config/templates.toml`. By exposing the empty string to the formatter, it's easier for users to override the strings if they like. That's what we do with the empty description placeholder, for example. The disadvantage is that users have to do that every time they use a custom template if they want a placeholder.

What you have here is more similar to the current behavior, so it should be less controversial. If we're unsure, I think we can merge this as is and decide later if we want to move it to the default template.

File: cli/tests/test_templater.rs [hunk 1]
```diff
@@ -418,6 +418,30 @@ fn test_templater_signature() {
     insta::assert_snapshot!(render(r#"author"#), @"Test User <x@y>");
     insta::assert_snapshot!(render(r#"author.email()"#), @"x@y");
     insta::assert_snapshot!(render(r#"author.username()"#), @"x");
+
+    let (_, stderr) = test_env.jj_cmd_ok(&repo_path, &["--config-toml=user.name=''", "new"]);
+    insta::assert_snapshot!(stderr, @r###"
```
Comment (martinvonz) [inline cli/tests/test_templater.rs:423]: I think this is covered in `test_global_opts.rs` (which I think was written before we had `--config-toml`, btw)

File: lib/src/settings.rs [hunk 2]
```diff
@@ -98,19 +98,14 @@ impl UserSettings {
...
-        self.config
-            .get_string("user.name")
-            .unwrap_or_else(|_| Self::user_name_placeholder().to_string())
+        self.config.get_string("user.name").unwrap_or_default()
     }
 
-    pub fn user_name_placeholder() -> &'static str {
-        "(no name configured)"
-    }
+    // Must not be changed to avoid git pushing older commits with no set name
```
Comment (martinvonz) [inline lib/src/settings.rs:104]: We typically drop support for old formats after 6 months or so, but most upgrades so far have automatically upgraded existing repo. In this case, I don't think we'll want to rewrite existing commits, so I guess we should keep support for these old commits much longer. It's not very costly in terms of maintenance anyway. I'm mostly thinking out loud, but let me know if you disagree.

**Review by martinvonz — APPROVED** [review]
Nice! And thanks for bringing this case up earlier. I like the new implementation much better.

File: lib/src/git_backend.rs [hunk 3]
```diff
@@ -241,9 +241,21 @@ fn commit_from_git_without_root_parent(commit: &git2::Commit) -> Commit {
     }
 }
 
+const EMPTY_STRING_PLACEHOLDER: &str = "JJ_EMPTY_STRING";
+
 fn signature_from_git(signature: git2::Signature) -> Signature {
-    let name = signature.name().unwrap_or("<no name>").to_owned();
-    let email = signature.email().unwrap_or("<no email>").to_owned();
+    let name = signature.name().unwrap_or(EMPTY_STRING_PLACEHOLDER);
```
[the remark this answers is not in the document]
Comment (emilykfox) [reply lib/src/git_backend.rs:247]: Done.

File: lib/src/settings.rs (hunk 2 above)
Comment (emilykfox) [reply lib/src/settings.rs:104]: I agree we should continue supporting the old placeholders. Rewriting commits(' hashes) would be very distributive.

File: cli/src/template_builder.rs
Comment (emilykfox) [reply cli/src/template_builder.rs:1]: I agree with merging for now. I created https://github.com/martinvonz/jj/issues/2112 for further discussion.

File: cli/tests/test_templater.rs (hunk 1 above)
Comment (emilykfox) [reply cli/tests/test_templater.rs]: I removed the `stderr` checks but kept the placeholder replacement checks.

**Pull request merged**